File: cpp/examples/bad_service.cpp
// ❌ BAD: This file demonstrates common C++ anti-patterns. Do NOT use as a template.

#include <cstdio>
#include <cstring>
#include <iostream>
#include <string>

// ❌ BAD: global mutable state — not thread-safe
const char* g_api_key = "sk_live_HARDCODED_KEY"; // ❌ BAD: secret in source
static int  g_order_count = 0;                    // ❌ BAD: global counter

// ❌ BAD: raw pointer ownership with manual new/delete — memory leak risk
struct Order {
    char* id;       // ❌ BAD: raw char* instead of std::string
    double total;   // ❌ BAD: double for money — floating-point errors
    char* status;   // ❌ BAD: raw char* for status instead of enum
};

// ❌ BAD: factory returns raw pointer — caller must remember to delete
Order* create_order(const char* id, double total) {
    Order* o = new Order(); // ❌ BAD: raw new — must be manually deleted
    o->id = strdup(id);     // ❌ BAD: strdup returns malloc'd memory, no RAII
    o->total = total;
    o->status = strdup("pending");
    g_order_count++;
    return o;               // ❌ BAD: ownership transfer unclear
}

// ❌ BAD: throws generic std::exception in hot path — no typed error info
// ❌ BAD: takes raw pointer — ownership semantics unclear
bool process_payment(Order* order, const char* token) {
    if (order == nullptr) {
        throw std::exception(); // ❌ BAD: too generic, no message
    }

    // ❌ BAD: strcmp for equality — error-prone, no null check
    if (strcmp(order->status, "pending") != 0) {
        std::cout << "not pending" << std::endl; // ❌ BAD: std::cout in library code
        return false;
    }

    // ❌ BAD: sprintf with user-supplied data — buffer overflow if id is long
    char query[64];
    sprintf(query, "SELECT * FROM orders WHERE id = '%s'", order->id); // ❌ BAD: SQL injection + overflow

    // ❌ BAD: printf for logging — no timestamp, no level, not structured
    printf("processing order %s\n", order->id);

    // ❌ BAD: using global API key
    printf("using key: %s\n", g_api_key);

    return true;
    // ❌ BAD: free(order->id) never called — memory leak
    // ❌ BAD: order is never deleted — memory leak
}

// ❌ BAD: C-style cast instead of static_cast / reinterpret_cast
void update_status(Order* order, void* new_status) {
    order->status = (char*)new_status; // ❌ BAD: C-style cast, aliasing issues
}

// ❌ BAD: No virtual destructor — UB when deleting derived through base pointer
class BaseService {
public:
    virtual void Process() {}
    ~BaseService() {} // ❌ BAD: must be virtual ~BaseService()
};

class DerivedService : public BaseService {
    int* resource = new int(42); // ❌ BAD: raw new in member, no destructor
public:
    void Process() override {}
    // ❌ BAD: no destructor — resource leaked when deleted via BaseService*
};

File: rust/examples/bad_service.rs
// ❌ BAD: This file demonstrates common Rust anti-patterns. Do NOT use as a template.
#![allow(dead_code, unused_variables)]

use std::collections::HashMap;

// ❌ BAD: global mutable static — not thread-safe without synchronisation
static mut API_KEY: &str = "sk_live_HARDCODED_KEY"; // ❌ BAD: secret in source

// ❌ BAD: String-typed status — not exhaustive, no compile-time safety
pub struct BadOrder {
    pub id: String,        // ❌ BAD: plain String, no newtype — confusable with user_id
    pub user_id: String,
    pub status: String,    // ❌ BAD: "pending", "paid", etc. as raw strings
    pub total: f64,        // ❌ BAD: f64 for money — floating-point rounding errors
}

// ❌ BAD: function returns Option<String> for errors — no context about what went wrong
pub fn process_payment(order_id: &str, token: &str) -> Option<String> {
    if order_id.is_empty() {
        return None; // ❌ BAD: None gives no error information
    }

    // ❌ BAD: unsafe block to access global mutable static — data race in multithreaded context
    let key = unsafe { API_KEY };

    // ❌ BAD: .unwrap() will panic if the vec is empty — no graceful error handling
    let orders: Vec<BadOrder> = Vec::new();
    let order = orders.iter().find(|o| o.id == order_id).unwrap(); // ❌ BAD: panics

    if order.status != "pending" {
        println!("order not pending"); // ❌ BAD: println! in library code
        return None;
    }

    // ❌ BAD: format! used to build a SQL query — SQL injection if order_id is user-supplied
    let query = format!("SELECT * FROM payments WHERE order_id = '{}'", order_id);

    // ❌ BAD: clone everywhere to avoid understanding ownership
    let token_copy = token.to_string().clone(); // ❌ BAD: redundant .to_string().clone()
    let id_copy = order_id.to_owned().clone();  // ❌ BAD: redundant .to_owned().clone()

    // ❌ BAD: Box<dyn std::error::Error> as return type on public API — opaque to callers
    // Should use thiserror or a concrete enum.

    Some("fake-payment-id".to_string())
}

// ❌ BAD: using String keys and dynamic dispatch (HashMap<String, Box<dyn Any>>) 
// instead of typed structs — loses compile-time guarantees
pub fn get_order_details(id: &str) -> HashMap<String, String> {
    let mut map = HashMap::new();
    map.insert("id".to_string(), id.to_string());
    // ❌ BAD: missing fields silently omitted — caller doesn't know what's present
    map
}

// ❌ BAD: mixing sync I/O inside an async fn — will block the executor thread
pub async fn fetch_and_cancel(order_id: &str) {
    // ❌ BAD: std::thread::sleep blocks the async executor, starving other tasks
    std::thread::sleep(std::time::Duration::from_secs(2));
    // Should use: tokio::time::sleep(Duration::from_secs(2)).await;
    println!("cancelled {}", order_id);
}
